src/parser/validator.ts

```diff
@@ -0,0 +1,52 @@
+import { RootParser } from "../parser/RootParser";
+import { makeCstVisitor, VisitorResult, CommandNode } from "../parser/makeCstVisitor";
+
+interface validateCSTResult {
+    errors: ValidateCSTError[];
+    hasErrors: boolean;
+}
+
+interface ValidateCSTError {
+    error: Error,
+}
+
+export function validateVisitorResult(visitorResult: VisitorResult): validateCSTResult {
+    var validateResult = { errors: [], hasErrors: false};
+    var idSet = new Set<String>();
+    visitorResult.commands.forEach((command) => {
+        validateCST(command, validateResult, idSet);
+    });
+    return validateResult;
+}
+
+function validateCST(commandNode: CommandNode, errors: validateCSTResult, idSet: Set<String>) {
+    commandNode.children.forEach((child) => {
+        validateCST(child, errors, idSet);
+    });
+    codeBlockHasID(commandNode, errors);
+    idIsUnique(commandNode, errors, idSet);
```

Comment:
This is a neat pattern idea, but it does rely on in/out parameters a bit much and it is confined to this function block. These validators are also tied to the specific commands.

It might be cool to define a "Rule" or "Validator" as a function that takes a command node and a context object (which would contain idSet, but we could extend it later) and returns an array of errors (empty if no errors). Then we can add validators to a list which we'll run through on a node with something like: 

```js
const validators = [idIsUnique, codeBlockHasId];
// ... for each command node ...
const allErrorsForNode = validators.reduce((acc, cur) => [...acc, ...cur(commandNode, {idSet})]);
```

Rather than having the validator check whether it is working on a specific command like "code-block", for example, we can provide a different list of validators to each command definition, so like:

```
const validators = {
  "code-block": [idIsUnique, codeBlockHasId],
  "something-else": [idIsUnique],
};
// ...
const allErrorsForNode = validators[commandNode.commandName].reduce(...);
```

This would be an elegant way to describe the requirements of each command and reuse validator code.

Reply:
Did *something* like this -- let me know if you think I should tweak it some more.